File: SFROpack/NM01/demon1bitCS.py
# Solving 1 bit compressive sensing using randomly generated data

from solver.examples.onebcs.func1BCS import func1BCS
from solver.examples.onebcs.refine import refine
from solver.NM01 import NM01
from solver.examples.onebcs.PlotRecovery import plot_recovery
from solver.examples.onebcs.random1bcs import datageneration1BCS

import numpy as np

n = 1000
m = int(np.ceil(0.5 * n))
s = int(np.ceil(0.01 * n))            # sparsity level
r = 0.01                              # flipping ratio
nf = 0.05                             # noisy ratio

A, c, co, xo = datageneration1BCS('Ind', m, n, s, nf, r, 0.5)  # data generation

func = lambda x, key: func1BCS(x, key, 1e-5, 0.5, A, c)
B = (-c[:, None]) * A
# Ensure b is 1-D to match B @ x shape (m,)
b = (n * 8e-5) * np.ones(m)
lam = 1

pars = {}
pars['tau'] = 1
pars['strict'] = (n <= 2000)

out = NM01(func, B, b, lam, pars)
x = refine(out['sol'], s, A, c)

plot_recovery(xo, x, pos=(950, 500, 500, 250), show_info=True)

print(' Computational time:    %.3fsec' % out['time'])
print(' Signal-to-noise ratio: %.2f' % (-20 * np.log10(np.linalg.norm(x - xo))))
print(' Hamming distance:      %.3f' % (np.count_nonzero(np.sign(A @ x) - c) / m))
print(' Hamming error:         %.3f' % (np.count_nonzero(np.sign(A @ x) - co) / m))

File: SFROpack/NM01/demon1bitCS.m
% Solving 1 bit compressive sensing using randomly generated data 
clc; close all; clear all; addpath(genpath(pwd)); %#ok<CLALL>


n            = 1000; 
m            = ceil(0.5*n);
s            = ceil(0.01*n);                      % sparsity level
r            = 0.01;                              % flipping ratio
nf           = 0.05;                              % noisy ratio
[A,c,co,xo]  = random1bcs('Ind',m,n,s,nf,r,0.5);  % data generation

func         = @(x,key)func1BCS(x,key,1e-5,0.5,A,c);
B            = (-c).*A;
b            = (n*8e-5)*ones(m,1);
lam          = 1;
pars.tau     = 1;  
pars.strict  =(n<=2000); 
out          = NM01(func, B, b, lam, pars); 
x            = refine(out.sol,s,A,c);

PlotRecovery(xo,x,[950,500,500,250],1)
fprintf(' Computational time:    %.3fsec\n',out.time);
fprintf(' Signal-to-noise ratio: %.2f\n',-20*log10(norm(x-xo)));
fprintf(' Hamming distance:      %.3f\n',nnz(sign(A*x)-c)/m)
fprintf(' Hamming error:         %.3f\n',nnz(sign(A*x)-co)/m)